packages/core/src/browser/hover-service.ts

```diff
@@ -0,0 +1,189 @@
+
+import '../../src/browser/style/hover-service.css';
+
+export type HoverPosition = 'left' | 'right' | 'top' | 'bottom';
+
+export namespace HoverPosition {
+    export function invertIfNecessary(position: HoverPosition, target: DOMRect, host: DOMRect, totalWidth: number, totalHeight: number): HoverPosition {
+        if (position === 'left') {
+            if (target.left - host.width - 5 < 0) {
+                return 'right';
+            }
+        } else if (position === 'right') {
+            if (target.right + host.width + 5 > totalWidth) {
+                return 'left';
+            }
+        } else if (position === 'top') {
+            if (target.top - host.height - 5 < 0) {
+                return 'bottom';
+            }
+        } else if (position === 'bottom') {
+            if (target.bottom + host.height + 5 > totalHeight) {
+                return 'top';
+            }
+        }
+        return position;
+    }
+}
+
+export interface HoverRequest {
+    content: string | MarkdownString | HTMLElement
+    target: HTMLElement
+    /**
+     * The position where the hover should appear.
+     * Note that the hover service will try to invert the position (i.e. right -> left)
+     * if the specified content does not fit in the window next to the target element
+     */
+    position: HoverPosition
+}
+
+@injectable()
+export class HoverService {
+    protected static hostClassName = 'theia-hover';
+    protected static styleSheetId = 'theia-hover-style';
+    @inject(PreferenceService) protected readonly preferences: PreferenceService;
+    @inject(MarkdownRendererFactory) protected readonly markdownRendererFactory: MarkdownRendererFactory;
+
+    protected _markdownRenderer: MarkdownRenderer | undefined;
+    protected get markdownRenderer(): MarkdownRenderer {
+        this._markdownRenderer ||= this.markdownRendererFactory();
+        return this._markdownRenderer;
+    }
+
+    protected _hoverHost: HTMLElement | undefined;
+    protected get hoverHost(): HTMLElement {
+        if (!this._hoverHost) {
+            this._hoverHost = document.createElement('div');
+            this._hoverHost.classList.add(HoverService.hostClassName);
+            this._hoverHost.style.position = 'absolute';
+        }
+        return this._hoverHost;
+    }
+    protected pendingTimeout: Disposable | undefined;
+    protected hoverTarget: HTMLElement | undefined;
+    protected lastHidHover = Date.now();
+    protected readonly disposeOnHide = new DisposableCollection();
+
+    requestHover(request: HoverRequest): void {
+        if (request.target !== this.hoverTarget) {
+            this.cancelHover();
+            this.pendingTimeout = disposableTimeout(() => this.renderHover(request), this.getHoverDelay());
+        }
+    }
+
+    protected getHoverDelay(): number {
+        return Date.now() - this.lastHidHover < 200
+            ? 0
+            : this.preferences.get('workbench.hover.delay', isOSX ? 1500 : 500);
+    }
```

Comment:
@msujew, I notice that in Theia with this change, if _any_ hover has recently been rendered, then any other hover, of the same or different kind, will be rendered immediately, but in VSCode, it appears to differ by source. So e.g. if you hover over an extension tree node, get the hover, then hover over a  sidebar tabbar icon, the tabbar icon's title hover will appear immediately, but in VSCode, it will only hover after a delay.